src/Http/ServerRequest.php

```diff
@@ -553,9 +553,12 @@ protected function _is(string $type, array $args): bool
     protected function _acceptHeaderDetector(array $detect): bool
     {
         $content = new ContentTypeNegotiation();
-        $accepted = $content->preferredType($this, $detect['accept']);
+        $options = $detect['accept'];
+        // We add text/html as the browsers often combine xml + html types together.
```

Comment:
I think this needs a clearer comment about the effect we want to have.

Reply:
In writing a better explanation I realized that this change would break detecting HTML types. I've change the implementation around a bit and used a more explicit ignore/exclude list instead.

Reply:
So we're saying if you have a request with html and xml accept headers and have only an xml renderer, then this isn't a valid xml request.

Reply:
> So we're saying if you have a request with html and xml accept headers and have only an xml renderer, then this isn't a valid xml request.

Kind of. It depends on the header value. An Accept header of  `text/html,application/xml` will not count as XML. As both XML and HTML are of equal preference value.

However, a request of `text/html;q=0.8, application/xml` *will* count as XML as HTML is a lower preference value. Given that many browsers include `text/html,application/xml;q=0.9` in their headers we can't only check if XML is accepted because while the browser will accept XML, it prefers HTML more.

Adding in a special case for the XML type seemed like the safest change to me. The previous implementation of `_headerDetector` was:

```php
    protected function _headerDetector(array $detect): bool
    {
        foreach ($detect['header'] as $header => $value) {
            $header = $this->getEnv('http_' . $header);
            if ($header !== null) {
                if (!is_string($value) && !is_bool($value) && is_callable($value)) {
                    return $value($header);
                }

                return $header === $value;
            }
        }

        return false;
    }
```

This logic was doing a strict check on only the unqualified types and `application/xml` != `application/xml;q=0.9`.

Reply:
Ok, that makes sense to me. We're marking exclusions at equal preference. Instead of writing a sorting order, we're special casing exclusions. Other requests with matching equality get what they get from the server.